File: ocaml/list-ops/list_ops.ml
let rec fold ~init ~f lst =
	match lst with
		[] -> init |
		x::xs ->
			fold ~init:(f init x) ~f:f xs

let reverse =
	fold ~init:[] ~f:(fun acc v -> v::acc)

let foldr ~init ~f lst =
	fold ~init:init ~f:f (reverse lst)

let length =
	fold ~init:0 ~f:(fun acc _ -> acc + 1)

let map ~f =
	let op' acc v = 
		(f v) :: acc
	in
		foldr ~init:[] ~f:op'

let filter ~f =
	let op' acc v =
		if f v then
			v::acc
		else
			acc
	in
		foldr ~init:[] ~f:op'

let append lst1 lst2 =
	fold ~init:(reverse lst1) ~f:(fun acc v -> v::acc) lst2
		|> reverse

let concat lst =
	lst
	(*foldr ~init:[] ~f:append*)

File: elixir/list-ops/list_ops.exs
defmodule ListOps do
  # Please don't use any external modules (especially List) in your
  # implementation. The point of this exercise is to create these basic functions
  # yourself.
  #
  # Note that `++` is a function from an external module (Kernel, which is
  # automatically imported) and so shouldn't be used either.

  @spec count(list) :: non_neg_integer
  def count(l) do
    case l do
      [head | rest] -> 1 + count(rest)
      [head]        -> 1
      []            -> 0
    end
  end

  @spec reverse(list) :: list
  def reverse(l) do
    reduce(l, [], fn (x, acc) -> [x | acc] end)
  end

  @spec map(list, (any -> any)) :: list
  def map(l, f) do
    l |> reverse |> reduce([], fn (x, acc) -> [f.(x) | acc] end)
  end

  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter(l, f) do
    processor = fn x, acc ->
      case f.(x) do
        true -> [ x | acc]
        false -> acc
      end
    end
    l |> reverse |> reduce([], processor)
  end

  @type acc :: any
  @spec reduce(list, acc, ((any, acc) -> acc)) :: acc
  def reduce(l, acc, f) do
    case l do
      [head | rest]  -> memo = f.(head, acc)
                        reduce(rest, memo, f)
      [head]         -> f.(head, acc)
      []             -> acc
    end
  end

  @spec append(list, list) :: list
  def append(a, b) do
    processor = fn x, acc ->
      [x | acc]
    end
    reduce(b, reverse(a), processor) |> reverse
  end

  @spec concat([[any]]) :: [any]
  def concat(ll) do
    reduce(ll, [], fn (l, acc) ->
      reduce(l, acc, fn (v, acc2) -> [v | acc2] end)
    end) |> reverse
    # Enum.flat_map(ll, fn(x) -> x end)
  end
end